### PR TITLE
Revise how staking configurations are set

### LINKED ISSUE #10938: Revised how staking configurations are set
Currently, whenever we want to set staking parameters via [`set_staking_config`](https://paritytech.github.io/substrate/master/pallet_staking/pallet/struct.Pallet.html#method.set_staking_configs), we need to name all of the previous values as well, and double check that we are not altering them. This is a bit of a hassle. Preferably, this transaction would wrap all of its arguments in a 

```rust
enum ConfigOp<T: Default + Codec> {
    /// Don't change.
    Noop,
    /// Set to the given value.
    Set(T),
    /// remove from storage.
    Remove, 
}
```

No storage migration should be needed.

NOTE: I am open to alternative ways of doing this as well. Using an enum adds a few bytes to the transactions, which is totally fine since this is not user-dispatchable. 

I considered wrapping all arguments in another option, and if set to `None` it would not set it, but that is too ambiguous in my opinion.

### PR DESCRIPTION
**Pull request by hirschenberger**
fixes #10938

polkadot companion: https://github.com/paritytech/polkadot/pull/5018

polkadot address: 1ZApdbwrNnT1Hey4xeTAH6SZPDgbA9ZVftNuu9jTRbkc2Bg

### CODE REVIEW

**Comment by hirschenberger** [comment]
We assume the worst case weight here but it may actually be much more lightweight. 
So it would be possible to refund some weight afterwards.

**Review by shawntabrizi — APPROVED** [review]
lgtm. needs to run weights again after merge

**Review by shawntabrizi — CHANGES_REQUESTED** [review]
On closer review, actually a few things needs to be patched up:

* You need to make sure all tests pass: `cargo test -p pallet-staking --features runtime-benchmarks`
* Having `ConfigOp<Option<...>>` does not make sense, since `ConfigOp` has all of `Set`, `Remove`, and `Noop`. In the case of `ConfigOp::set(None)`, you probably want to actually do `ConfigOp::Remove`, so really there should be no `Option`s.
* To support options in storage, you can use the `put` api instead of `set` api.
* Add a test that the `put` behavior is working as expected with the underlying storage.
* Look at the tests that you fixed up, and introduce the usage of `ConfigOp::Noop` or `ConfigOp::Remove` where appropriate. For example, if the value does not change from one point to the next.

But the general idea is good, and the way you did weights is good enough imo.

File: frame/staking/src/pallet/mod.rs [hunk 1]
```diff
@@ -60,6 +61,16 @@ pub mod pallet {
 	#[pallet::without_storage_info]
 	pub struct Pallet<T>(_);
 
+	#[derive(TypeInfo, Debug, Clone, Encode, Decode, PartialEq)]
```
Comment (kianenigma) [inline frame/staking/src/pallet/mod.rs:64]:
```suggestion
	/// Possible operations on the configuration values of this pallet. 	
	#[derive(TypeInfo, Debug, Clone, Encode, Decode, PartialEq)]
```

File: frame/staking/src/tests.rs [hunk 2]
```diff
@@ -40,6 +40,62 @@ use sp_staking::{
...
+			ConfigOp::Noop,
+			ConfigOp::Noop,
+			ConfigOp::Noop
+		));
+		assert_eq!(MinNominatorBond::<Test>::get(), 1_500);
+		assert_eq!(MinValidatorBond::<Test>::get(), 2_000);
+		assert_eq!(MaxNominatorsCount::<Test>::get(), Some(10));
+		assert_eq!(MaxValidatorsCount::<Test>::get(), Some(20));
+		assert_eq!(ChillThreshold::<Test>::get(), Some(Percent::from_percent(75)));
+		assert_eq!(MinCommission::<Test>::get(), Perbill::from_percent(0));
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:78]:
```suggestion
		assert_storage_noop!(Staking::set_staking_configs(
			Origin::root(),
			ConfigOp::Noop,
			ConfigOp::Noop,
			ConfigOp::Noop,
			ConfigOp::Noop,
			ConfigOp::Noop,
			ConfigOp::Noop
		));
```

**Review by kianenigma — APPROVED** [review]
With @shawntabrizi's feedback now it looks good to me! Thanks 💯

**Review by shawntabrizi — APPROVED** [review]

**Comment by shawntabrizi** [comment]
@hirschenberger are you able to make a companion PR, or do you need help with that?

I would just copy and paste `set_staking_configs` benchmarks, and paste it twice with the appropriate names.

**Comment by hirschenberger** [comment]
> @hirschenberger are you able to make a companion PR, or do you need help with that?
> 
> I would just copy and paste `set_staking_configs` benchmarks, and paste it twice with the appropriate names.

I can do a companion

**Review by shawntabrizi — APPROVED** [review]

**Comment by kianenigma** [comment]
bot merge

**Comment by kianenigma** [comment]
bot merge

**Pull request merged**